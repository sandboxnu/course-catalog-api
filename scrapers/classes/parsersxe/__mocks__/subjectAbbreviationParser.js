import data from "./subjectAbbreviationTable.json";

export function getSubjectAbbreviations() {
  return data;
}
