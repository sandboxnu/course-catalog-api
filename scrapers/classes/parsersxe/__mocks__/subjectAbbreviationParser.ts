import data from "./subjectAbbreviationTable.json";

export function getSubjectAbbreviations(): Record<string, string> {
  return data;
}
