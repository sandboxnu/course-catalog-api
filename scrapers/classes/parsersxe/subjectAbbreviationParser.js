/* eslint-disable @typescript-eslint/no-use-before-define */
/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import _ from "lodash";
import he from "he";
import macros from "../../../utils/macros";
import Request from "../../request";

const request = new Request("subjectAbberviationParser");

/**
 * Get the subject abberviations for use in parsing prereqs
 */
export const getSubjectAbbreviations = _.memoize(async (termId) => {
  macros.log(
    `SubjectAbberviationParser: Not memoized. Scraping term ${termId}`
  );
  const subjectResponse = await requestSubjects(termId);
  return createDescriptionTable(subjectResponse);
});

export const getSubjectDescriptions = _.memoize(async (termId) => {
  const subjectResponse = await requestSubjects(termId);
  return createAbbrTable(subjectResponse);
});

async function requestSubjects(termId) {
  const MAX = 500; // If there are more than 500 THIS WILL BREAK. Would make it smarter but not worth it rn.
  const URL =
    "https://nubanner.neu.edu/StudentRegistrationSsb/ssb/courseSearch/get_subject";
  const subjectUrl = `${URL}?searchTerm=&term=${termId}&offset=1&max=${MAX}`;
  const response = await request.get({
    url: subjectUrl,
    json: true,
  });
  if (response.statusCode !== 200) {
    macros.error(`Problem with request for subjects ${subjectUrl}`);
  }
  return response.body;
}

function createDescriptionTable(subjects) {
  subjects = subjects.map((subject) => {
    return {
      subjectCode: subject.code,
      description: he.decode(subject.description),
    };
  });
  subjects = _.keyBy(subjects, "description");
  return _.mapValues(subjects, "subjectCode");
}

function createAbbrTable(subjects) {
  subjects = subjects.map((subject) => {
    return {
      description: he.decode(subject.description),
      subjectCode: subject.code,
    };
  });
  subjects = _.keyBy(subjects, "subjectCode");
  return _.mapValues(subjects, "description");
}

// Export for testing https://philipwalton.com/articles/how-to-unit-test-private-functions-in-javascript/
export { createDescriptionTable as _createDescriptionTable };
