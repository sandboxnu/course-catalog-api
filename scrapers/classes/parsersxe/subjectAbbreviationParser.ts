/* eslint-disable @typescript-eslint/no-use-before-define */
/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import _ from "lodash";
import he from "he";
import macros from "../../../utils/macros";
import Request from "../../request";
import { SubjectDescription } from "../../../types/scraperTypes";

const request = new Request("subjectAbberviationParser");

/**
 * Get the subject abberviations for use in parsing prereqs
 */
export const getSubjectAbbreviations = _.memoize(async (termId: string) => {
  macros.log(
    `SubjectAbbreviationParser: Not memoized. Scraping term ${termId}`
  );
  const subjectResponse = await requestSubjects(termId);
  return createDescriptionTable(subjectResponse);
});

export const getSubjectDescriptions = _.memoize(async (termId: string) => {
  const subjectResponse = await requestSubjects(termId);
  return createAbbrTable(subjectResponse);
});

async function requestSubjects(termId: string): Promise<SubjectDescription[]> {
  const MAX = 500; // If there are more than 500 THIS WILL BREAK. Would make it smarter but not worth it rn.
  const URL =
    "https://nubanner.neu.edu/StudentRegistrationSsb/ssb/courseSearch/get_subject";
  const subjectUrl = `${URL}?searchTerm=&term=${termId}&offset=1&max=${MAX}`;
  const response = await request.get({
    url: subjectUrl,
    json: true,
  });

  if (response.statusCode !== 200) {
    macros.error(`Problem with request for subjects ${subjectUrl}`);
  }
  return response.body;
}

function createDescriptionTable(
  subjects: SubjectDescription[]
): Record<string, string> {
  const mappedSubjects = subjects.map((subject) => {
    return {
      subjectCode: subject.code,
      description: he.decode(subject.description),
    };
  });

  const mappedByDesc = _.keyBy(mappedSubjects, "description");
  return _.mapValues(mappedByDesc, "subjectCode");
}

function createAbbrTable(
  subjects: SubjectDescription[]
): Record<string, string> {
  const mappedSubjects = subjects.map((subject) => {
    return {
      description: he.decode(subject.description) as string,
      subjectCode: subject.code,
    };
  });

  const mappedByCode = _.keyBy(mappedSubjects, "subjectCode");
  return _.mapValues(mappedByCode, "description");
}

// Export for testing https://philipwalton.com/articles/how-to-unit-test-private-functions-in-javascript/
export { createDescriptionTable as _createDescriptionTable };
