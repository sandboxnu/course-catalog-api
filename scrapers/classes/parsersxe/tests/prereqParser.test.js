import PrereqParser from "../prereqParser";
import PrereqParserData from "./data/prereqParser.data";
import { getSubjectAbbreviations } from "../subjectAbbreviationParser";

jest.mock("../subjectAbbreviationParser");
const subjectAbbreviationTable = getSubjectAbbreviations();

describe("prereqParser", () => {
  it("should parse corequisites", () => {
    const phys1161coreqs = PrereqParser.serializeCoreqs(
      PrereqParserData.coreqs.phys1161,
      subjectAbbreviationTable
    );
    const phys1161actual = {
      type: "and",
      values: [
        { classId: "1162", subject: "PHYS" },
        { classId: "1163", subject: "PHYS" },
      ],
    };
    const hlth1201coreqs = PrereqParser.serializeCoreqs(
      PrereqParserData.coreqs.hlth1201,
      subjectAbbreviationTable
    );
    const hlth1201actual = {
      type: "and",
      values: [{ classId: "1200", subject: "HLTH" }],
    };
    expect(phys1161coreqs).toEqual(phys1161actual);
    expect(hlth1201coreqs).toEqual(hlth1201actual);
  });

  it("should parse prerequisites", () => {
    const prereqs = PrereqParser.serializePrereqs(
      PrereqParserData.prereqs.biol3405,
      subjectAbbreviationTable
    );
    const actual = {
      type: "or",
      values: [
        { classId: "1103", subject: "BIOL" },
        { classId: "1113", subject: "BIOL" },
        { classId: "2297", subject: "BIOL" },
        { classId: "2299", subject: "BIOL" },
        { classId: "2290", subject: "ENVR" },
        { classId: "2290", subject: "EEMB" },
        { classId: "3458", subject: "PSYC" },
      ],
    };
    expect(prereqs).toEqual(actual);
  });

  it("shouldn't break if there are no prerequisites/corequisites", () => {
    const empty = {
      type: "and",
      values: [],
    };
    expect(
      PrereqParser.serializeCoreqs(PrereqParserData.coreqs.biol3405)
    ).toEqual(empty);
    expect(
      PrereqParser.serializePrereqs(PrereqParserData.prereqs.cs2500)
    ).toEqual(empty);
  });

  it('should add "Graduate Admissions REQ" as a string', () => {
    const prereqs = PrereqParser.serializePrereqs(
      PrereqParserData.prereqs.biol5549,
      subjectAbbreviationTable
    );
    const actual = {
      type: "or",
      values: [{ subject: "BIOL", classId: "2301" }, "Graduate Admission"],
    };
    expect(prereqs).toEqual(actual);
  });

  it("should handle parenthesized prerequisites", () => {
    const prereqs = PrereqParser.serializePrereqs(
      PrereqParserData.prereqs.chem5610,
      subjectAbbreviationTable
    );
    const actual = {
      type: "or",
      values: [
        {
          type: "and",
          values: [
            {
              type: "or",
              values: [
                { classId: "2317", subject: "CHEM" },
                { classId: "2313", subject: "CHEM" },
              ],
            },
            {
              type: "or",
              values: [
                { classId: "3401", subject: "CHEM" },
                { classId: "3421", subject: "CHEM" },
                { classId: "3431", subject: "CHEM" },
              ],
            },
          ],
        },
        "Graduate Admission",
      ],
    };
    expect(prereqs).toEqual(actual);
  });

  it("should handle parenthesized prerequisites 2", () => {
    const prereqs = PrereqParser.serializePrereqs(
      PrereqParserData.prereqs.cs4240,
      subjectAbbreviationTable
    );
    expect(prereqs).toMatchSnapshot();
  });

  it('should handle nested parenthesized prerequisites or "Graduate Admissions REQ"', () => {
    const cive2221prereqs = PrereqParser.serializePrereqs(
      PrereqParserData.prereqs.cive2221,
      subjectAbbreviationTable
    );
    const nrsg2220prereqs = PrereqParser.serializePrereqs(
      PrereqParserData.prereqs.nrsg2220,
      subjectAbbreviationTable
    );
    expect(cive2221prereqs).toMatchSnapshot();
    expect(nrsg2220prereqs).toMatchSnapshot();
  });
});
