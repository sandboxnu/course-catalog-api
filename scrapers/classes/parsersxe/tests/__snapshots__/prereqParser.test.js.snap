// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prereqParser should handle nested parenthesized prerequisites or "Graduate Admissions REQ" 1`] = `
{
  "type": "or",
  "values": [
    {
      "type": "and",
      "values": [
        {
          "classId": "1342",
          "subject": "MATH",
        },
        {
          "type": "or",
          "values": [
            {
              "classId": "1151",
              "subject": "PHYS",
            },
            {
              "classId": "1161",
              "subject": "PHYS",
            },
          ],
        },
      ],
    },
    {
      "type": "and",
      "values": [
        {
          "classId": "1141",
          "subject": "PHYS",
        },
        {
          "classId": "1241",
          "subject": "MATH",
        },
      ],
    },
  ],
}
`;

exports[`prereqParser should handle nested parenthesized prerequisites or "Graduate Admissions REQ" 2`] = `
{
  "type": "or",
  "values": [
    {
      "type": "and",
      "values": [
        {
          "classId": "1119",
          "subject": "BIOL",
        },
        {
          "classId": "1121",
          "subject": "BIOL",
        },
        {
          "type": "or",
          "values": [
            {
              "classId": "1101",
              "subject": "CHEM",
            },
            {
              "classId": "1211",
              "subject": "CHEM",
            },
          ],
        },
      ],
    },
    "Graduate Admission",
  ],
}
`;

exports[`prereqParser should handle parenthesized prerequisites 2 1`] = `
{
  "type": "and",
  "values": [
    {
      "type": "or",
      "values": [
        {
          "classId": "3600",
          "subject": "CS",
        },
        {
          "classId": "3650",
          "subject": "CS",
        },
        {
          "classId": "5600",
          "subject": "CS",
        },
        {
          "classId": "5600",
          "subject": "CS",
        },
      ],
    },
    {
      "type": "or",
      "values": [
        {
          "classId": "4800",
          "subject": "CS",
        },
        {
          "classId": "5800",
          "subject": "CS",
        },
        {
          "classId": "5800",
          "subject": "CS",
        },
      ],
    },
  ],
}
`;
